cupy/cuda/stream.pyx

```diff
@@ -182,16 +193,18 @@ class BaseStream(object):
 
     def __enter__(self):
         tls = _ThreadLocal.get()
-        if tls.prev_stream_ref_stack is None:
-            tls.prev_stream_ref_stack = []
+        cdef int dev = runtime.getDevice()
+        if tls.prev_stream_ref_stack[dev] is None:
+            tls.prev_stream_ref_stack[dev] = []
         prev_stream_ref = tls.get_current_stream_ref()
-        tls.prev_stream_ref_stack.append(prev_stream_ref)
+        tls.prev_stream_ref_stack[dev].append(prev_stream_ref)
         tls.set_current_stream(self)
         return self
 
     def __exit__(self, *args):
         tls = _ThreadLocal.get()
-        prev_stream_ref = tls.prev_stream_ref_stack.pop()
+        cdef int dev = runtime.getDevice()
```

Comment:
Should we need to keep the device id on `__enter__` as the current device may be changed in the block?

Reply:
I have been thinking about this too. It is more complicated than I expected. I feel it is still necessary to tie a stream to a device so we know which device the stream was created on without looking up the current device (which may not be the same as you pointed out).

Reply:
Hmmm...I changed my mind and now feel this should be illegal:
```python
with cuda.Device(0):
    stream0 = cuda.Stream()
    with stream0:
        with cuda.Device(1):
            ...  # <--- stream0 does not work on device 1
```

Reply:
It has to be legal so that we can switch device, create a new stream, and set it current.